OptimizelySDK/Bucketing/DecisionService.cs

```diff
@@ -70,61 +70,63 @@ public DecisionService(Bucketer bucketer, IErrorHandler errorHandler, UserProfil
         public virtual Result<Variation> GetVariation(Experiment experiment,
-            string userId,
-            ProjectConfig config,
-            UserAttributes filteredAttributes)
+            OptimizelyUserContext user,
+            ProjectConfig config)
         {
-            return GetVariation(experiment, userId, config, filteredAttributes, new OptimizelyDecideOption[] { });
+            return GetVariation(experiment, user, config, new OptimizelyDecideOption[] { });
         }
 
         /// <summary>
         /// Get a Variation of an Experiment for a user to be allocated into.
         /// </summary>
         /// <param name = "experiment" > The Experiment the user will be bucketed into.</param>
-        /// <param name = "userId" > The userId of the user.
-        /// <param name = "filteredAttributes" > The user's attributes. This should be filtered to just attributes in the Datafile.</param>
+        /// <param name = "user" > optimizely user context.
+        /// <param name = "config" > Project Config.</param>
+        /// <param name = "options" >An array of decision options.</param>
         /// <returns>The Variation the user is allocated into.</returns>
         public virtual Result<Variation> GetVariation(Experiment experiment,
-            string userId,
+            OptimizelyUserContext user,
             ProjectConfig config,
-            UserAttributes filteredAttributes,
             OptimizelyDecideOption[] options)
         {
             var reasons = new DecisionReasons();
+            var userId = user.GetUserId();
             if (!ExperimentUtils.IsExperimentActive(experiment, Logger))
                 return Result<Variation>.NullResult(reasons);
 
             // check if a forced variation is set
-            var forcedVariationResult = GetForcedVariation(experiment.Key, userId, config);
-            reasons += forcedVariationResult.DecisionReasons;
+            var decisionVariation = GetForcedVariation(experiment.Key, userId, config);
+            reasons += decisionVariation.DecisionReasons;
+            var variation = decisionVariation?.ResultObject;
 
-            if (forcedVariationResult.ResultObject != null)
+            if (variation == null)
             {
-                return forcedVariationResult.SetReasons(reasons);
-            }
+                decisionVariation = GetWhitelistedVariation(experiment, user.GetUserId());
+                reasons += decisionVariation?.DecisionReasons;
 
-            var variationResult = GetWhitelistedVariation(experiment, userId);
-            reasons += variationResult.DecisionReasons;
+                variation = decisionVariation?.ResultObject;
+            }
 
-            if (variationResult.ResultObject != null)
-            {                
-                return variationResult.SetReasons(reasons);
+            if (variation != null)
+            {
+                return decisionVariation;
```

Comment:
We'll lose reasons updated in line 120 and above

Reply:
Not sure what does it mean but decisionVariation has all reasons set above in line 120.